modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js

```diff
@@ -0,0 +1,386 @@
+							) {
+								direction = CKEDITOR.SELECTION_BOTTOM_TO_TOP;
+							}
+						}
+
+						return direction;
+					},
+				});
+
+				CKEDITOR.ui.balloonPanel.prototype.attach = (function () {
+					return function (elementOrSelection, options) {
+						var padding =
+							CKEDITOR.ui.balloonPanel.DEFAULT_PANEL_PADDING;
+
+						if (
+							options instanceof CKEDITOR.dom.element ||
+							!options
+						) {
+							options = {focusElement: options};
+						}
+
+						options = CKEDITOR.tools.extend(options, {
+							show: true,
+						});
+
+						if (options.show === true) {
+							this.show();
+						}
+
+						this.fire('attach');
+
+						var editable = this.editor.editable();
+						var editableClientRect = editable.getClientRect(true);
+						var panelClientRect = this.parts.panel.getClientRect(
+							true
+						);
+						var type = elementOrSelection.getType();
+						var x = 0;
+						var y = 0;
+
+						if (type === CKEDITOR.SELECTION_TEXT) {
+							var direction = this._getSelectionDirection();
+							var ranges = elementOrSelection.getRanges();
+							var range = ranges[0];
+							var rangeClientRects = range.getClientRects(true);
+
+							var firstClientRect = rangeClientRects[0];
+							var lastClientRect =
+								rangeClientRects[rangeClientRects.length - 1];
+
+							if (firstClientRect === lastClientRect) {
+								direction = CKEDITOR.SELECTION_BOTTOM_TO_TOP;
+							}
+
+							var rangeHeight =
+								lastClientRect.bottom - firstClientRect.top;
+							var clientRect = rangeClientRects[0];
+
+							x =
+								clientRect.x +
+								clientRect.width / 2 -
+								panelClientRect.width / 2;
+							y = clientRect.y - panelClientRect.height - padding;
+
+							if (
+								direction === CKEDITOR.SELECTION_TOP_TO_BOTTOM
+							) {
+								y = clientRect.y + rangeHeight + padding;
+							}
+						}
+						else if (type === CKEDITOR.SELECTION_ELEMENT) {
+							var selectedElement = elementOrSelection.getSelectedElement();
+
+							if (!selectedElement) {
+								ranges = elementOrSelection.getRanges();
+								selectedElement =
+									ranges && ranges[0].startContainer;
+							}
+
+							var selectedElementClientRect = selectedElement.getClientRect(
+								true
+							);
+
+							// Center toolbar horizontally and on top of the selected element
+
+							x =
+								selectedElementClientRect.x +
+								selectedElementClientRect.width / 2 -
+								panelClientRect.width / 2;
+							y =
+								selectedElementClientRect.y -
+								panelClientRect.height -
+								padding;
+						}
+
+						if (x < editableClientRect.x) {
+							x = editableClientRect.x + padding;
+						}
+						if (
+							x + panelClientRect.width >
+							editableClientRect.width
+						) {
+							x -=
+								x +
+								panelClientRect.width -
+								editableClientRect.width;
+						}
+
+						if (y < editableClientRect.y) {
+							y = clientRect.bottom + padding;
```

Comment:
```suggestion
							y = editableClientRect.bottom + padding;
```

It fixes an issue when trying to move the image around the editor :D

Reply:
From: https://liferay.slack.com/archives/C3JBR21HA/p1622644861109600?thread_ts=1622562932.067900&cid=C3JBR21HA

It solves:

**Errors when positioning an image** point **2**

**Misc** points **1,2 and 3**

Reply:
Hey @diegonvs, please be explicit about the error even when including a link to a Slack thread (_Misc points 1,2 and 3_ is pretty vague even when viewing the thread) - I'll have a look at the suggested solution, but I don't understand how this could affect "pasting text" and the "image drag handle".

Reply:
Hey again @diegonvs,

We should discuss this, because I'm not seeing the same issues you mention

> 1. If I select a very large text and copy and just paste it throws some errors

In this .gif, I'm pasting 15 paragraphs of "Lorem Ipsum" a couple of times (note that the issues are related to cookies and not the editor, no error is thrown)

![recording (3)](https://user-images.githubusercontent.com/5572/120603265-1db6a600-c44c-11eb-87c9-a928bc4f40c3.gif)


> 2. If I select a very large text and try to format something it throws some errors

In this .gif, I select 15 paragraphs of text and formatting it to bold, underline, italic and the removing the formatting: no errors are thrown

![](https://user-images.githubusercontent.com/5572/120602436-4e4a1000-c44b-11eb-80d2-106dfed857c3.gif)


> 3. Looks like the image dragging handle is not working

In this .gif, I'm resizing the image a few times using the "handle", as I said in the thread "it's not pixel perfect", but it's not so bad either. This plugin comes with CKEditor, we haven't done any modifications so far.

![recording (4)](https://user-images.githubusercontent.com/5572/120603977-dbda2f80-c44c-11eb-95bb-f4de749390d4.gif)

I'm not trying to say that there are no errors, I'm not seeing what you described so if I didn't understand what you wrote, please include a .gif and the error itself. Thanks

Reply:
**[EDIT 2]** To avoid this being "committed" since I don't agree, I'm deleting your suggestion

![2021-06-03-09:35:54](https://user-images.githubusercontent.com/5572/120606272-4c824b80-c44f-11eb-93ce-6e47352794e0.png)